### ISSUE TITLE
Set up and teardown for individual tests

### ISSUE THREAD

**Issue by hackebrot** [mozilla/redash-ui-tests#3]
We need a way to run tests in isolation that they don't leak into other test items. The order in which tests are being executed must not impact the test results.

- [x] pytest setup to create clean database tables etc.
- [ ] pytest teardown to drop database tables etc.

**Comment by hackebrot** [comment]
Check out [redash_client](https://github.com/mozilla/redash_client)!

**Comment by hackebrot** [comment]
Only what we need to log in a user.

**Comment by jrbenny35** [comment]
> Only what we need to log in a user.

So kind of like what we have now? I just want to see if I need to clean up what we have now, or do something different.

**Comment by hackebrot** [comment]
Since we don't have access to the ``Factory`` that is being used in the redash unit tests, we need to create test data manually. For the **login test** we need an organization and a user, see https://github.com/mozilla/redash-ui-tests/blob/master/tests/conftest.py#L58

**Comment by jrbenny35** [comment]
Why can't we create test data before we run the tests? Currently on [addons-server](https://github.com/mozilla/addons-server/blob/master/tox.ini#L58) we create the database and fill it with data, and then run the tests. It helps with using the tests in an environment outside of the one we can control, say a dev or stage instance.

**Comment by hackebrot** [comment]
We can do that, sure. For some tests we will have to clean the data that we created during the test execution. I'm not entirely sure I understand what you're asking. Are you suggesting to use make rather than pytest fixtures for the setup?

**Comment by jrbenny35** [comment]
I think that is what I am saying. I did some research on how to use the python postgres client, and it would be more work than I think it is worth. We are really at a loss here not being able to use upstream redash features. If we can somehow call the commands within the redash container, that would help a bunch. But I don't think we want to rewrite every sql command that we need by hand,

**Comment by jrbenny35** [comment]
Per chat:

Initial Setup: Create DB, maybe setup root user.
Per-Test: Ability to add/remove data from database as needed.

**Comment by jrbenny35** [comment]
Per IRC
```
I suggest to have commands in Makefile/CI to
1. Create the database
2. Create the database tables
3. Create root user

And a pytest fixture to 
1. create all users from variables.json

And tests that
1. log in a user who was created in the pytest fixture
```

**Comment by jrbenny35** [comment]
I reached out to Jannis and he pointed me [here](https://github.com/getredash/redash/blob/master/redash/handlers/users.py#L28-L61). I've got this working locally.

**Comment by jezdez** [comment]
@jrbenny35 There is a [`users` CLI subcommand](https://github.com/getredash/redash/blob/672c86efaa4eee7ed4295fe021431409e418de57/redash/cli/users.py) for the `manage.py` script, that can be used to [delete a user](https://github.com/getredash/redash/blob/672c86efaa4eee7ed4295fe021431409e418de57/redash/cli/users.py#L147-L166): `./manage.py users delete foo@bar.com`

Could you use that instead?

**Comment by jezdez** [comment]
@jrbenny35 Right, since the API doesn't support out of the box this I'm afraid there is only one option left to add new endpoint ourselves using the fairly new [Redash extensions mechanism](https://github.com/getredash/redash/pull/2354). It's the way we're going to move our Redash customizations into a separate package called [redash-stmo](https://github.com/mozilla/redash-stmo) as well.

It's a bit more involved but here are the steps to do that:

1. Create a custom Python package, e.g. redash_ui_tests, that follows the same API as [redash-stmo](https://github.com/mozilla/redash-stmo)
2. To that package add a Redash extension that registers a new custom `UserDeleteResource` class that does the deletion like you want with the passed Flask app instance similar to how the [built-in `UserDisableResource` is registered](https://github.com/getredash/redash/blob/617df8e4f65f2f7bbf904b385043f46f980cdcd5/redash/handlers/api.py#L94) under the endpoint `/api/users/<user_id>/delete`.
3. Release the package to PyPI
4. Install it in the Docker image by adding it to the `Pipfile`
5. Call to the custom user delete API endpoint in your teardown code

**Comment by jrbenny35** [comment]
@jezdez that is a great solution I will have to look into! Thanks for that

**Comment by jezdez** [comment]
@jrbenny35 The great thing about it is that you can extend it with other API endpoints or code changes if needed and keep it in sync with this repo. I guess that package can easily just live in this repo next to the tests to simplify coordination. Just put it into `src/redash_ui_tests` and add the `setup.py` etc files.

**Comment by hackebrot** [comment]
Hi @jrbenny35 @jezdez! 👋 

Can we create a story for this over at [getredash/redash](https://github.com/getredash/redash) or [mozilla/redash](https://github.com/mozilla/redash), as this seems like a missing capability of Redash rather than testing specific requirement?

Keep in mind that we can't install Redash extensions into a hosted Redash instance. So if we want to be able to run the UI tests against staging or a release candidate we need to work with the official API.

For the two login tests that we have, we don't need to be able to delete users, so we can go ahead without adding this at this point.

**Comment by jrbenny35** [comment]
I agree. It seems like the delete issue has been open for over 2 years. Makes sense if you are able to create you should be able to delete.

**Comment by jezdez** [comment]
@hackebrot There is already https://github.com/getredash/redash/issues/799 and a related issue was fixed recently in https://github.com/getredash/redash/pull/1600. Not sure if upstream will add an API endpoint to delete users since right now the frontend client is the only consumer of that API. Since we're working on a new consumer with there ui tests I would keep it contained here until there is an indication of interest to implement it in the main repo.

The long term goal to support any Redash instance shouldn't be a blocker to implement our user deletion endpoint now to get this off the ground. For our instance of Redash we already install redash-stmo in stage and prod and it's perfectly fine to assume that we can install redash-ui-tests there as well.

**Comment by hackebrot** [comment]
The goal to support both Redash running inside of Docker container or on a hosted environment is not a blocker to implement such an endpoint. However I feel like this belongs in [redash-stmo](https://github.com/mozilla/redash-stmo) rather than the UI testing harness.

I'd rather move forward on getting the login tests up and running, than spending time to create a new PyPI package with an endpoint that is not strictly required at this point.

**Comment by jezdez** [comment]
@hackebrot Makes sense, let's table it until we really need it.

**Comment by jrbenny35** [comment]
@hackebrot so for now just get the user creation working? I kinda hate half-implementing stuff but I can understand as there isn't a solution as of right now.

**Comment by jezdez** [comment]
@hackebrot @jrbenny35 Let's open a ticket in https://github.com/mozilla/redash-stmo and link to this tread, ok?

**Comment by hackebrot** [comment]
Don't worry about that, @jrbenny35! Since we're not testing the creation of users via the GUI, there's not need for us to remove the users at this point. Our goal is to get the login tests working! 👍

**Comment by jrbenny35** [comment]
Solved the first one in #26

**Comment by hackebrot** [comment]
Closing this now. Let's open new issues for specific test cases!
